From 41f163494615bec36f5672fb15034ad6e40ffc91 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Wed, 11 Oct 2017 15:55:00 +0200
Subject: [PATCH] OS-6398 support bhyve as a platform Reviewed by: Patrick
 Mooney <patrick.mooney@joyent.com> Reviewed by: Robert Mustacchi
 <rm@joyent.com> Approved by: Robert Mustacchi <rm@joyent.com>

---
 usr/src/uts/i86pc/io/pcplusmp/apic_common.c | 4 +++-
 usr/src/uts/i86pc/os/cpuid.c                | 4 ++++
 usr/src/uts/intel/sys/x86_archext.h         | 5 ++++-
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/usr/src/uts/i86pc/io/pcplusmp/apic_common.c b/usr/src/uts/i86pc/io/pcplusmp/apic_common.c
index 3892bc3f85..f4dbcf5dcf 100644
--- a/usr/src/uts/i86pc/io/pcplusmp/apic_common.c
+++ b/usr/src/uts/i86pc/io/pcplusmp/apic_common.c
@@ -1566,6 +1566,7 @@ apic_check_msi_support()
 	dev_info_t *cdip;
 	char dev_type[16];
 	int dev_len;
+	int hwenv = get_hwenv();
 
 	DDI_INTR_IMPLDBG((CE_CONT, "apic_check_msi_support:\n"));
 
@@ -1587,7 +1588,8 @@ apic_check_msi_support()
 			continue;
 		if (strcmp(dev_type, "pciex") == 0)
 			return (PSM_SUCCESS);
-		if (strcmp(dev_type, "pci") == 0 && get_hwenv() == HW_KVM)
+		if (strcmp(dev_type, "pci") == 0 &&
+		    (hwenv == HW_KVM || hwenv == HW_BHYVE))
 			return (PSM_SUCCESS);
 	}
 
diff --git a/usr/src/uts/i86pc/os/cpuid.c b/usr/src/uts/i86pc/os/cpuid.c
index daaa56ff5e..f44c75acd2 100644
--- a/usr/src/uts/i86pc/os/cpuid.c
+++ b/usr/src/uts/i86pc/os/cpuid.c
@@ -696,6 +696,10 @@ determine_platform(void)
 			platform_type = HW_KVM;
 			return;
 		}
+		if (strcmp(hvstr, HVSIG_BHYVE) == 0) {
+			platform_type = HW_BHYVE;
+			return;
+		}
 		if (strcmp(hvstr, HVSIG_MICROSOFT) == 0)
 			platform_type = HW_MICROSOFT;
 	} else {
diff --git a/usr/src/uts/intel/sys/x86_archext.h b/usr/src/uts/intel/sys/x86_archext.h
index 646aa5ac81..c06f3098ef 100644
--- a/usr/src/uts/intel/sys/x86_archext.h
+++ b/usr/src/uts/intel/sys/x86_archext.h
@@ -898,6 +898,7 @@ extern void xsave_setup_msr(struct cpu *);
 #define	HVSIG_VMWARE	"VMwareVMware"
 #define	HVSIG_KVM	"KVMKVMKVM"
 #define	HVSIG_MICROSOFT	"Microsoft Hv"
+#define	HVSIG_BHYVE	"bhyve bhyve "
 
 /*
  * Defined hardware environments
@@ -909,8 +910,10 @@ extern void xsave_setup_msr(struct cpu *);
 #define	HW_VMWARE	(1 << 3)	/* Running on VMware hypervisor */
 #define	HW_KVM		(1 << 4)	/* Running on KVM hypervisor */
 #define	HW_MICROSOFT	(1 << 5)	/* Running on Microsoft hypervisor */
+#define	HW_BHYVE	(1 << 6)	/* Running on bhyve hypervisor */
 
-#define	HW_VIRTUAL	(HW_XEN_HVM | HW_VMWARE | HW_KVM | HW_MICROSOFT)
+#define	HW_VIRTUAL	(HW_XEN_HVM | HW_VMWARE | HW_KVM | HW_MICROSOFT | \
+	    HW_BHYVE)
 
 #endif	/* _KERNEL */
 
-- 
2.21.0

